File: nodejs/hash.js
// this hash version requires the When library for dealing with async events

var W = require("when");
var nodefn = require("when/node/function");
var fs = require("fs");
var crypto = require("crypto");

var CHUNK_SIZE = 64 * 1024;

module.exports = function(path) {
  var buffer = new Buffer(CHUNK_SIZE * 2);

  return nodefn.call(fs.stat, path).then(function(stat) {
    var fileSize = stat.size;

    return nodefn.call(fs.open, path, "r").then(function(fd) {
      var reads = [0, fileSize - CHUNK_SIZE].map(function(offset, index) {
        return nodefn.call(fs.read, fd, buffer, CHUNK_SIZE * index, CHUNK_SIZE, offset);
      });

      return W.all(reads).then(function() {
        var md5sum = crypto.createHash("md5");
        md5sum.update(buffer);

        return md5sum.digest("hex");
      });
    });
  });
};

File: go/hash.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"io"
	"log"
	"os"
)

const chunkSize = 64 * 1024

func Hash(path string) (string, error) {
	hash := md5.New()

	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()

	if _, err := io.CopyN(hash, file, chunkSize); err != nil {
		return "", err
	}
	if _, err := file.Seek(-chunkSize, os.SEEK_END); err != nil {
		return "", err
	}
	if _, err := io.CopyN(hash, file, chunkSize); err != nil {
		return "", err
	}

	return hex.EncodeToString(hash.Sum(nil)), nil
}

func main() {
	log.SetFlags(0)
	for _, arg := range os.Args[1:] {
		hash, err := Hash(arg)
		if err != nil {
			log.Fatalln(err)
		}
		println(hash)
	}
}
